parlai/core/teachers.py

```diff
@@ -1422,6 +1427,140 @@ def _setup_data(self, path):
+    handle file parsing for you.
+
+    The data should be set up so that each dialogue instance (or, episode) 
+    occupies one line of valid JSON. The way the data is set up is as follows 
+    (with line breaks for readability):
+
+    ::
+
+        {
+            'dialogue':[
+                {'id':'modelx', 'text': 'hi'},
+                {'id':'modely', 'text': 'hi back'},
+                ...
+            ]
+        }
+
+    Note that by default, dialogs are interpreted as being one-way. 
+    For example, consider this dialog:
+
+    ::
+
+        {
+            'dialogue':[
+                {'id':'modelx', 'text': X1},
+                {'id':'modely', 'text': Y1},
+                {'id':'modelx', 'text': X2},
+                {'id':'modely', 'text': Y2},
+                {'id':'modelx', 'text': X3},
+                {'id':'modely', 'text': Y3},
+            ]
+        }
+
+    A set of examples X1 => Y1, X2 => Y2, and X3 => Y3 will be generated,
+    forming one episode. However, Y1 => X2 and Y2 => X3 are not created as 
+    separate examples by default. 
+    To change this behavior, you can set opt['label_turns']. The default 
+    value is 'secondspeaker' (i.e., the second speaker's utterances are
+    used as labels), but 'firstspeaker' and 'both' are also options. In the
+    case of 'both', two episodes are generated for each conversation.
+    """
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        if not shared:
+            self.episodes = []
+            self.num_exs = 0
+            self.label_turns = opt.get('label_turns')
+            if opt.get('conversationteacher_datafile') is not None:
+                self._setup_data(opt.get('conversationteacher_datafile'))
+        else:
+            self.episodes = shared['episodes']
+            self.num_exs = sum(len(e) for e in self.episodes)
+
+        self.id = opt['task']
+
+        self.reset()
+
+    def share(self):
+        """
+        Share the episodes.
+        """
+        shared = super().share()
+        shared['episodes'] = self.episodes
+        return shared
+
+    def num_examples(self):
+        """
+        Return the number of examples from the data.
+        """
+        return self.num_exs
+
+    def num_episodes(self):
+        """
+        Return the number of episodes from the data.
+        """
+        return len(self.episodes)
+
+    def get(self, episode_idx, entry_idx=None):
+        """
+        Get a specific example from the dataset.
+        """
+        return Message(self.episodes[episode_idx][entry_idx])
+
+    def _setup_data(self, path):
+        logging.info("[loading data from json file into task:" + path + "]")
+        self.episodes = []
+        self.num_exs = 0
+        eps = []
+        conversations = Conversations(path)
+        self.num_exs = sum(len(c) for c in conversations)
+        for conv in conversations:
+            context = conv.context
```

Comment:
I wasn't sure what to do with the context @emilydinan @stephenroller ? 
If you look at the sample json file I used for testing, the context got stored within the dialogue under the id 'context' so the dialogue doesn't really use it correctly.